introduce bs_version option for bs4 compatibility

madflow opened the pull request:
- [x] unit tests (we use [Jasmine 2.x.x](https://jasmine.github.io/2.2/introduction))
- [x] JSFiddle (or an equivalent such as CodePen, Plunker, etc) or screenshot/GIF with new feature or bug-fix
- [x] Link to original Github issue (if this is a bug-fix)
- [x] documentation updates to README file
- [ ] examples within [/tpl/index.tpl](https://github.com/seiyria/bootstrap-slider/blob/master/tpl/index.tpl) (for new options being added)
- [x] Passes CI-server checks (runs automated tests, JS source-code linting, etc..). To run these on your local machine, type `grunt test` in your Terminal within the bootstrap-slider repository directory

----

This is a first stab at Bootstrap 4 compatibilty. The idea here: Introduce a `bs_version` (`bs3` is default) (`data-slider-bs-version="bs4|bs3"`) option in order to simply support both upstream versions (3+4) in the same branch. It simply deals with new `Tooltip` css classes introduced in BS4. It is somewhat hacky :S 

The tooltips work - but they seem to need some more positioning fixes...

Fixes: https://github.com/seiyria/bootstrap-slider/issues/689
Plunkr with bs4 hard-coded as default: https://plnkr.co/edit/6fyg2tpFhNdKHQdgZ757

I have not added an example to the `tpl/index.tpl` because I am unsure how it would be best to:

* Install both versions of Boostrap as dev dependencies and have a sane build pipeline
* Deal with no "Less" files in the Bootstrap 4 package

seiyria commented on README.md:
```diff
@@ -199,6 +199,7 @@ Options can be passed either as a data (data-slider-foo) attribute, or as part o
 | focus | bool | false | Focus the appropriate slider handle after a value change. |
 | labelledby | string,array | null | ARIA labels for the slider handle's, Use array for multiple values in a range slider. |
 | rangeHighlights | array | [] | Defines a range array that you want to highlight, for example: [{'start':val1, 'end': val2, 'class': 'optionalAdditionalClassName'}]. |
+| bs_version | string | 'bs3' | Switch between Boostrap 3 ('bs3') and Bootstrap 4 ('bs4') |
```
camelcase

seiyria commented:
I don't know if this is _entirely_ accurate (@rovolution can correct me), but we don't want bootstrap as a dependency in any way. you can add the relevant tooltip classes to our css files. this will solve both of your queries.

Thanks for taking the time to tackle this (this is one of our most requested features).

I've left a small comment on the implementation. I also think it would be best if you just add a "bs3" or "bs4" class to the slider and then let the rest of the classes sort themselves out on their own - that will make this change much simpler overall.

madflow commented:
Maybe could elaborate what is meant by "but we don't want bootstrap as a dependency in any way" ?

If this is true the current codebase should already be functioning without Bootstrap (which is not the case when it comes to the Tooltips at least I believe).

If  the `bootstrap-slider` should just become `generic-slider` - then we do not need bs3 and bs4 altogether ;)

seiyria commented:
All I mean is that we copy the styles into bootstrap slider instead of having a hard dependency on them.

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -647,14 +647,15 @@ const windowIsDefined = (typeof window === "object");
 				}, this);
 
 				// Undo inline styles and classes on tooltips
+				var positionClasses = this._getTooltipPositionClassMap();
```
could you make this a `const` instead?

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1124,23 +1126,28 @@ const windowIsDefined = (typeof window === "object");
+				return  this.options.bs_version === 'bs4' ? 'show' : 'in';	
+			},
 			_showTooltip: function() {
+				var showClass = this._getTooltipShowClass();
```
could you make this a const instead?

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1124,23 +1126,28 @@ const windowIsDefined = (typeof window === "object");
 				this._state.over = true;
 			},
 			_hideTooltip: function() {
+				var showClass = this._getTooltipShowClass();
```
could you make this a const instead?

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1371,27 +1378,27 @@ const windowIsDefined = (typeof window === "object");
 					var offset_min = this.tooltip_min.getBoundingClientRect();
 					var offset_max = this.tooltip_max.getBoundingClientRect();
-
+					var positionClasses = this._getTooltipPositionClassMap();
```
could you make this a const instead?

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1851,7 +1858,17 @@ const windowIsDefined = (typeof window === "object");
 				return this.options.scale.toPercentage.apply(this, [value]);
 			},
+			_getTooltipPositionClassMap() {
+				var prefix = this.options.bs_version === 'bs4' ? 'tooltip-' : '';
```
could you make this a const instead?

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1851,7 +1858,17 @@ const windowIsDefined = (typeof window === "object");
+				};
+			},
 			_setTooltipPosition: function(){
+				var positionClasses = this._getTooltipPositionClassMap();
```
could you make this a const instead?

seiyria commented:
Yes, in that case do that. I couldn't remember but I'll defer to that.

On Thu, Feb 22, 2018, 19:43 Rohit Kalkur <notifications@github.com> wrote:

> *@rovolution* requested changes on this pull request.
>
> Thank you for this PR! As @seiyria <https://github.com/seiyria>
> mentioned, this has been high on our to-do list, so thank you for taking on
> the effort to do this!
>
> Personally, I do think we should be maintaining the dependency on
> bootstrap CSS, simply because this project is intended to be used with
> Bootstrap styles as we currently indicate this in our setup instructions
> <https://github.com/seiyria/bootstrap-slider#basic-setup>
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144735>
> :
>
> > @@ -647,14 +647,15 @@ const windowIsDefined = (typeof window === "object");
>  				}, this);
>
>  				// Undo inline styles and classes on tooltips
> +				var positionClasses = this._getTooltipPositionClassMap();
>
> could you make this a const instead?
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144757>
> :
>
> >  			_showTooltip: function() {
> +				var showClass = this._getTooltipShowClass();
>
> could you make this a const instead?
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144811>
> :
>
> >  					this.tooltip.style.display = 'none';
>  				}
>  				this._state.over = true;
>  			},
>  			_hideTooltip: function() {
> +				var showClass = this._getTooltipShowClass();
>
> could you make this a const instead?
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144824>
> :
>
> > @@ -1371,27 +1378,27 @@ const windowIsDefined = (typeof window === "object");
>
>  					var offset_min = this.tooltip_min.getBoundingClientRect();
>  					var offset_max = this.tooltip_max.getBoundingClientRect();
> -
> +					var positionClasses = this._getTooltipPositionClassMap();
>
> could you make this a const instead?
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144837>
> :
>
> > @@ -1851,7 +1858,17 @@ const windowIsDefined = (typeof window === "object");
>  			_toPercentage: function(value) {
>  				return this.options.scale.toPercentage.apply(this, [value]);
>  			},
> +			_getTooltipPositionClassMap() {
> +				var prefix = this.options.bs_version === 'bs4' ? 'tooltip-' : '';
>
> could you make this a const instead?
> ------------------------------
>
> In src/js/bootstrap-slider.js
> <https://github.com/seiyria/bootstrap-slider/pull/819#discussion_r170144858>
> :
>
> >  			_setTooltipPosition: function(){
> +				var positionClasses = this._getTooltipPositionClassMap();
>
> could you make this a const instead?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seiyria/bootstrap-slider/pull/819#pullrequestreview-98788683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAum2cY61M2omd6NhbcX-dem_MuRK81pks5tXhfXgaJpZM4SPO0J>
> .
>

madflow commented:
I rebased and implemented the changes. Let me know if there is anything missing on my end.

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1851,7 +1858,17 @@ const windowIsDefined = (typeof window === "object");
 				return this.options.scale.toPercentage.apply(this, [value]);
 			},
+			_getTooltipPositionClassMap() {
+				const prefix = this.options.bsVersion === 'bs4' ? 'tooltip-' : '';
```
could you wrap this conditional in parenthesis? makes the code a bit more readable

```js
const prefix = (this.options.bsVersion === 'bs4') ? 'tooltip-' : '';
```

rovolution commented on src/js/bootstrap-slider.js:
```diff
@@ -1124,23 +1126,28 @@ const windowIsDefined = (typeof window === "object");
 				}
 			},
+			_getTooltipShowClass: function() {
+				return  this.options.bsVersion === 'bs4' ? 'show' : 'in';	
```
same feedback: https://github.com/seiyria/bootstrap-slider/pull/819/commits/555e7fcc789bc430fcecbf80235cf4d0a01b2bfb#r172343511

madflow commented:
I implemented the changes and rebased. However: I thing I will have to revisit how tooltips work in Bootstrap 4: https://getbootstrap.com/docs/4.0/components/tooltips/

I think the correct Bootstrap class would be "bs-tooltip-{placement}" (I misread or botched it, dunno) and the third party "popper.js" should be included when we will go down the "Bootstrap 4" way.

AlwaysVbNet commented:
Any news on the compatibility issue? Any idea when the compatible version will be released?

seiyria commented:
As no commits have been made, nothing has changed yet.

madflow commented:
I did a short revisit - the Bootstrap Tooltips prepend a "bs-tooltip-" to the placement . I amended the PR for this. 

The biggest issue I am facing is the positioning. I can not really point out what has changed and this exceeds my hopes for "quick win" with this PR.  Sorry for creating this PR prematuraly.  

At the moment I would argue to close this. It only provides the bare minimum: that the tooltips show up...

madflow commented:
I created a plunkr with my latest code for illustration purposes: 

https://plnkr.co/edit/kYjQC196w46gf7WFVNjR?p=preview

madflow commented:
This should be superseeded by https://github.com/seiyria/bootstrap-slider/pull/856
